### PR TITLE
Created the robust race conditions checker.

### LINKED ISSUE #13186: TestHarness should figure out and run prereqs when using regex and/or failed-tests
## Reason

Right now if a restart-type test fails in the test harness. There's no good way to rerun just that test with the `--failed-tests` flag. If you attempt to do so, the failed test will be skipped because the dependency won't be thrown into the test DAG.

## Design

When running a subset of tests it would be very handy to automatically find and run the prereqs of the sub-selected tests.

## Impact

Minor: Usability

### PR DESCRIPTION
**Pull request by milljm**
- Added files to create and test the race condition checker.
- Edited current files to make use of the race condition checker.

closes #13186

Rebased work done by @Hysrok on PR #13474

### CODE REVIEW

File: python/contrib/dag/__init__.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 from copy import copy, deepcopy
 from collections import deque
+import os
```
Comment (milljm) [inline python/contrib/dag/__init__.py]: We probably don't need this

**Review by permcody — DISMISSED** [review]
Really nice work @Hysrok! Congratulations on your first contribution to MOOSE.

**Comment by permcody** [comment]
Also thanks @milljm for getting this over the line. I've added a commit to fixup the existing race conditions. There are more that trigger in recover mode but those will require me to touch files that have been annotated for SQA so I'll get those later.

**Review by permcody — APPROVED** [review]

**Comment by permcody** [comment]
@milljm -

To replicate the error that I see with this PR try the following:

```
cd moose/test/tests/outputs/recover
run_tests -j4   # should run just fine
run_tests -j4 -d # run the new "pedantic check" - also works fine
run_tests -j4 --recover # still good now we are doing recover tests - note the skipped tests here!
run_tests -j4 -d --recover # ought oh.... All hell breaks loose somehow the recover=false parameter isn't being honored in the check!
```

**Pull request merged**